trace_decoder/src/imp.rs

```diff
@@ -0,0 +1,641 @@
+                                path.into_hash()
+                                    .context("invalid path length in direct state trie")?,
+                                rlp::decode(&bytes)
+                                    .context("invalid AccountRlp in direct state trie")?,
+                            )?;
+                        }
+                        mpt_trie::trie_ops::ValOrHash::Hash(h) => {
+                            acc.insert_hash_by_key(path, h)?;
+                        }
+                    };
+                    anyhow::Ok(acc)
+                },
+            )?;
+            let storage = storage
+                .into_iter()
+                .map(|(k, SeparateTriePreImage::Direct(v))| {
+                    v.items()
+                        .try_fold(StorageTrie::default(), |mut acc, (nibbles, hash_or_val)| {
+                            let path = TrieKey::from_nibbles(nibbles);
+                            match hash_or_val {
+                                mpt_trie::trie_ops::ValOrHash::Val(value) => {
+                                    acc.insert(path, value)?;
+                                }
+                                mpt_trie::trie_ops::ValOrHash::Hash(h) => {
+                                    acc.insert_hash(path, h)?;
+                                }
+                            };
+                            anyhow::Ok(acc)
+                        })
+                        .map(|v| (k, v))
+                })
+                .collect::<Result<_, _>>()?;
+            (state, storage, Hash2Code::new())
+        }
+        BlockTraceTriePreImages::Combined(CombinedPreImages { compact }) => {
+            let instructions = crate::wire::parse(&compact)
+                .context("couldn't parse instructions from binary format")?;
+            let crate::type1::Frontend {
+                state,
+                storage,
+                code,
+            } = crate::type1::frontend(instructions)?;
+            (state, storage, code.into_iter().map(Into::into).collect())
+        }
+    })
+}
+
+/// Break `txns` into batches of length `hint`, prioritising creating at least
+/// two batches.
+///
+/// [`None`] represents a dummy transaction that should not increment the
+/// transaction index.
+fn batch(txns: Vec<TxnInfo>, hint: usize) -> Vec<Vec<Option<TxnInfo>>> {
+    let hint = cmp::max(hint, 1);
+    let mut txns = txns.into_iter().map(Some).collect::<Vec<_>>();
+    let n_batches = txns.iter().chunks(hint).into_iter().count();
+    match (txns.len(), n_batches) {
+        // enough
+        (_, 2..) => txns
+            .into_iter()
+            .chunks(hint)
+            .into_iter()
+            .map(FromIterator::from_iter)
+            .collect(),
+        // not enough batches at `hint`, but enough real transactions
+        (2.., ..2) => {
+            let second = txns.split_off(txns.len() / 2);
+            vec![txns, second]
```

Comment:
we shouldn't need to do this if the line above `let hint = cmp::max(hint, 1);` was already taking into account the nb of txns FWIW (as is done in the current code), which then reads a bit better to me

Reply:
I don't think I'm smart enough to understand the old code :S

Reply:
I guess it's just that I find the approach with `chunk(batch_size)` cleaner / more readable.